### DISCUSSION TITLE
Navier-Stokes Preconditioning for a large problem - move across from mailing list

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#16113] [Q&A Modules: Navier-Stokes] [answered]
Hi 

So quick reminder I'm trying to solve a big (by my standards) fluid problem - its 2.4e6 elements - 16.6e6 DOF on 1280 cores, with a pre split mesh. Here is the run dump

```
Building mesh .                                                                            [  2.22 s]
Initializing equation system .............................                                 [ 30.02 s]

Framework Information:
MOOSE Version:           git commit 95fb74e5fe on 2020-10-01
LibMesh Version:         b51a1fac743ce11a1e65018f25a478241c24ed44
PETSc Version:           3.13.3
SLEPc Version:           3.13.3
Current Time:            Fri Nov  6 12:44:48 2020
Executable Timestamp:    Thu Oct  1 18:19:58 2020

Parallelism:
  Num Processors:          1280
  Num Threads:             1

Mesh: 
  Parallel Type:           distributed (pre-split)
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 3336156
    Local:                 3176
  Elems:                   
    Total:                 2406640
    Local:                 1935
  Num Subdomains:          1
  Num Partitions:          1
  Partitioner:             parmetis

Nonlinear System:
  AD size required:        90
  Num DOFs:                16680780
  Num Local DOFs:          15880
  Variables:               "velocity" { "p" "temp_fluid" } 
  Finite Element Types:    "LAGRANGE_VEC" "LAGRANGE" 
  Approximation Orders:    "SECOND" "SECOND" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             PostprocessorDT
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    FSP

Projecting initial condition .............................................................
........................................                                                   [102.88 s]
Computing initial stateful property values ......................                          [ 23.01 s]
Copying soultions back ...................................................................
..........................................................................................
........................                                                                   [182.53 s]

Time Step 0, time = 0
Warning:  This MeshOutput subclass only supports meshes which have been serialized!
Warning:  This MeshOutput subclass only supports meshes which have been serialized!

Postprocessor Values:
+----------------+----------------+
| time           | cfl            |
+----------------+----------------+
|   0.000000e+00 |   6.165143e+11 |
+----------------+----------------+

Outputting out 
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................................................................................
..........................                                                                 [477.73 s]

Time Step 1, time = 0.01, dt = 0.01
Computing initial residual .....                                                           [  6.41 s]

Performing automatic scaling calculation

 0 Nonlinear |R| = 1.948466e+04
Building node to element map .....................................................         [ 54.57 s]
      0 Linear |R| = 1.948466e+04
```
It's been at that stage for about 10 hours now, maybe even more like 15. When I ran this problem on 640 cores I got an error from petsc (error 9) suggesting a memory problem.

So my question is, any advice, different preconditioners? 

Thanks

See input attached - hmm downside of GH Discussion, can't attach arbitrary files. But here is the input as a txt, also some screens of the geometry.
[fluid_sim-pc-no-t.txt](https://github.com/idaholab/moose/files/5504696/fluid_sim-pc-no-t.txt)
![Screenshot from 2020-11-06 12-16-37](https://user-images.githubusercontent.com/2439754/98440557-60e8aa80-20f1-11eb-9968-5cc0ebbdbe88.png)
![Screenshot from 2020-11-06 12-16-56](https://user-images.githubusercontent.com/2439754/98440558-61814100-20f1-11eb-9cb8-aba810a3d50f.png)

**Comment by lindsayad** [comment]
So one thing straightaway that’s not related to your question... please see #16091 and #16089. You should ‘./configure —with-ad-indexing-type=global’ to avoid all that time spent in initial condition and “stateful property” computation. 

To your question: I have seen the FSP options that you’re using be fairly expensive in terms of memory and even CPU for large enough problems. I’ve found asm or block Jacobi with sub block LU to be the most effective preconditioner for large INS problems. But I do believe that with some investigation, working with @fdkong , we can get FSP to be successful

**Reply by fdkong** [reply]
Yes, we will keep exploring FSP options. We might eventually come up with some good built-in options at the end.

**Comment by makeclean** [comment]
Ah thats useful, I'll rebuild - there a whole bunch of options that are really useful, I wonder if they should be default?

So I'm now on;
```
  petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -sub_pc_factor_levels  -ksp_gmres_restart'
  petsc_options_value = 'asm      2               ilu          4       200'
```
That fails with out of memory, 
```
Time Step 1, time = 0.01, dt = 0.01
Computing initial residual .....                                                           [  6.28 s]

Performing automatic scaling calculation

 0 Nonlinear |R| = 1.948466e+04

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 27176 RUNNING AT arm02.ib0.diracarm.le.ac.uk
=   EXIT CODE: 9
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
```
Im going to try Jacobi next

**Reply by lindsayad** [reply]
This actually looks like it could be running out of memory during Jacobian assembly since we haven’t yet printed the first linear residual to the screen. I really want to get parallel heap profiling going (#15990). How big of a derivative vector did you configure with?

Something I can say is that you’re doing yourself no favors memory-wise by using second order basis. At least it looks like you’re using tets. Still, you will have rows with 50 coupled dofs (5 vars * 10 dofs per elem) instead of potentially 20 coupled dofs (4 dofs per elem). So that has the potential to reduce your memory footprint if you go to first order (you are already using PSPG). If you have significant curvature I would suggest just adding more elements to try and capture it unless that’s untenable for some reason. 

If you do go to first order you can reconfigure ‘—with-derivative-size=20’ to try and speed up your calculations just a bit (and potentially further reduce your memory footprint a smidge).

**Reply by makeclean** [reply]
I'm pretty sure I have derivative-size=180 currently I had a big structural problem which warranted it earlier and I've stuck with it since. 

I can try first order, and I have no issue with adding more elements I'll give that a bash too. Thus far (perhaps for historical reasons) I've tended to use 2nd order for velocity and 1st for pressure, perhaps with the ADINS stuff its no-longer necessary.

**Reply by lindsayad** [reply]
The mixed order basis is required for [LBB stability](https://en.wikipedia.org/wiki/Ladyzhenskaya%E2%80%93Babu%C5%A1ka%E2%80%93Brezzi_condition) if you are not using PSPG. PSPG makes equal-order basis stable. If you ever see a mixed-order basis do better with nonlinear/linear convergence than a PSPG equal-order basis, however, then I would be very interested to hear about it.

**Reply by fdkong** [reply]
I think you should use sparse AD that uses much less memory and also way faster (at least for one project I have done last month).

**Comment by makeclean** [comment]
Ok, with 2nd order as before but with 
```
  petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -sub_pc_factor_levels  -ksp_gmres_restart'
  petsc_options_value = 'bjacobi      2               ilu          4       200'
```
We now have lift off, well its running at least ;)
```
    300 Linear |R| = 1.091458e+02
  Linear solve did not converge due to DIVERGED_ITS iterations 300
 1 Nonlinear |R| = 2.581837e+06
      0 Linear |R| = 2.581837e+06
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC_FAILED due to FACTOR_NOERROR 
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 1
 Solve Did NOT Converge!
Aborting as solve did not converge

Time Step 1, time = 0.005, dt = 0.005
Computing initial residual .....                                                           [  6.79 s]
 0 Nonlinear |R| = 3.913081e+04
```
We'll see in the morning if it gets anywhere - thanks for the help so far

**Reply by makeclean** [reply]
Ok, so over night much more progress was made, we're getting through some linear and non-linear iterations now and we're failing to converge, but ive got the timestep set pretty high. I've seen the in past that it needs to take a very small timestep, say 1e-6 seconds, then its happy to take great big 1e-2 size time steps, so this is great progress!

**Reply by fdkong** [reply]
ILU(4) is quite expensive, and it might already very close to LU. I was wondering if the algorithm will still converge if you switch to use  ILU(3).

ASM uses more memory than bjacobi because of overlapping elements and also duplicated local submatrices

**Comment by makeclean** [comment, marked as answer]
So good news, I changed 1st order, and its flying now :) Taking great big honking time steps which is great, but im wondering if i should tighten up my tolerances?  

In terms of iterations im seeing a decent drop in linear residuals and in around 3 or 4 non-linear steps which is great, however... im seeing some weird velocity and temperature excursions,

![Screenshot from 2020-11-09 11-54-08](https://user-images.githubusercontent.com/2439754/98538298-70e3c400-2282-11eb-9014-b2650092a77c.png)
So this one shows temp which should be between 300 and 673 K, the walls are at 673 and its injected at 573, but the IC is 300, is this a question of more elements, but the temp should definitely not be greater than 637 or lower than 300 (given some interpolation between nodes) but certainly not in the -1e6 to 1e6 range.


![Screenshot from 2020-11-09 10-53-37](https://user-images.githubusercontent.com/2439754/98538495-be603100-2282-11eb-9db6-6ed694d9457b.png)
Velocity has some similar problems, we're ramping with time (max(1,t)) and I can forgive velocities in the range 1,4 probably, but there are local excursions to 5e3 m/s 

Does that indicate that I should lengthen up tolerances or mesh resolution?

**Reply by lindsayad** [reply]
I would try to avoid using `nl_abs_tol` unless you're stepping towards a steady-state. You may try using `nl_rel_tol = 1e-8`. But I think the more important part is probably to increase the mesh resolution to try to avoid oscillations in the fields (overshoots and undershoots). What is your Reynolds number? Also I can't quite tell in the temperature visualization with the red patch...is the majority of the domain at 300K? Or is most of the domain below 300 K? I would be much more concerned if we are seeing a majority of the domain go below a physical value as opposed to in some small regions near sharp discontinuities, which continuous FEM is notoriously bad at resolving. SUPG helps, but it's not a silver bullet.

**Comment by fdkong** [comment]
In general, adaptive time stepper may be helpful here

```
  [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 0.5
    growth_factor = 1.2
    cutback_factor = 0.8
  [../]
```

**Comment by mlesueur** [comment]
Hi @makeclean,
I have just seen your post as I was doing my annual errand on moose-users about: "is there a better preconditioner for my large simulations of Stokes flow" :grin: So I'm jumping on the post.
To give you some context, I am simulating pressure-driven Stokes flow through complex structures (rock microCT-scans) and I am often dealing with DOFs >1e7.
I have already discussed preconditioning with @fdkong and the one I am using right now is:

```javascript
[FSP]
    type = FSP
    petsc_options_iname = '-snes_type -ksp_type -ksp_rtol -ksp_atol -ksp_max_it -snes_atol -snes_rtol -snes_max_it -snes_max_funcs'
    petsc_options_value = 'newtonls     fgmres     1e-2     1e-15       200       1e-10        1e-15       200           100000'
    topsplit = 'uv'
    [uv]
      petsc_options_iname = '-pc_fieldsplit_schur_fact_type -pc_fieldsplit_schur_precondition'
      petsc_options_value = 'upper selfp'
      splitting = 'u v'
      splitting_type = schur
    []
    [u]
      vars = 'vel_x vel_y vel_z'
      petsc_options_iname = '-pc_type -ksp_type -pc_hypre_type'
      petsc_options_value = '  hypre    preonly     boomeramg '
    []
    [v]
      vars = 'p'
      petsc_options_iname = '-pc_type -ksp_type -sub_pc_type -sub_pc_factor_levels'
      petsc_options_value = '  jacobi  preonly        ilu            3'
    []
  []
```
It shows great performance and I am able to solve 1e7+ DOFs within an hour on 50 processors (also I'm using PSPG with first order elements). Let me know if this works for you!
Still for me the problem I am facing was mentioned by @lindsayad and @fdkong: the memory usage is off the charts! For example for 1e7 DOFs, it needs 64Gb (even with mesh splitting).
I would love to have any feedback on whether I can improve this FSP.
Otherwise I tested also the solutions mentioned here. It is the asm ilu that seemed to give me some good convergence with a low memory usage. HOWEVER, I have a a DIVERGED_LINE_SEARCH error after the second nonlinear iteration, see below. I have tried all the different line search options, or disabling it, but it does not work. Is it maybe because of the 0 in the diagonal of the stokes matrix? Do you think there is a solution to still have the asm ilu working?
Thank you in advance for your help,
Cheers,

Martin

```javascript
Computing initial residual                                                                 [  1.29 s]
 0 Nonlinear |R| = 1.976305e-04
      0 Linear |R| = 1.976305e-04
...
     26 Linear |R| = 1.017750e-07
 1 Nonlinear |R| = 9.125860e-05
      0 Linear |R| = 9.125860e-05
      1 Linear |R| = 9.125821e-05
      2 Linear |R| = 9.125519e-05
      3 Linear |R| = 9.125061e-05
      4 Linear |R| = 9.123911e-05
      5 Linear |R| = 9.122871e-05
      6 Linear |R| = 9.115078e-05
      7 Linear |R| = 9.111489e-05
      8 Linear |R| = 9.091965e-05
      9 Linear |R| = 9.079105e-05
     10 Linear |R| = 8.967217e-05
     11 Linear |R| = 8.818694e-05
     12 Linear |R| = 7.786348e-05
     13 Linear |R| = 6.424927e-05
     14 Linear |R| = 3.950608e-05
     15 Linear |R| = 2.839524e-05
     16 Linear |R| = 1.836606e-05
     17 Linear |R| = 1.196996e-05
     18 Linear |R| = 6.749037e-06
     19 Linear |R| = 3.362408e-06
     20 Linear |R| = 1.857514e-06
     21 Linear |R| = 1.028440e-06
     22 Linear |R| = 6.480960e-07
     23 Linear |R| = 3.930797e-07
     24 Linear |R| = 2.569072e-07
     25 Linear |R| = 1.634042e-07
     26 Linear |R| = 1.104569e-07
     27 Linear |R| = 6.666340e-08
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 1
 Solve Did NOT Converge!
Aborting as solve did not converge

Outlier Variable Residual Norms:
  vel_x: 8.300333e-05
```

**Reply by lindsayad** [reply]
High memory use with FSP...I do not see `lu` anywhere. @fdkong why would it be taking so much memory relative to `asm + ilu`?

Are you using PJFNK or NEWTON? could you run with `-ksp_monitor_true_residual` and paste in that same solve history? Also are you using the `INSAD` or `INS` kernels? Finally, you should not have a zero for any of the diagonals if you are using PSPG.

**Reply by fdkong** [reply]
In general, I do not expect the `FSP` memory usage is way more than  that of `ASM+ILU`. 

 It would be really helpful if @mlesueur   would like to do a heap profiling. With a profiling result, we might have an idea if we are able to improve anything. 

Here are some profiling instructions:  https://mooseframework.inl.gov/application_development/profiling.html

**Reply by fdkong** [reply]
> Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 1
 Solve Did NOT Converge!
Aborting as solve did not converge

 @mlesueur You could turn off line search, and try it again. Add `line_search = none` into executioner block

**Reply by lindsayad** [reply]
I think @fdkong means to be tagging @mlesueur here. It would be great to see a heap profile for the FSP case (and also for the asm-ilu case for comparison).

**Reply by makeclean** [reply]
either way I appreciate the ping to remind me of my old problem :)

**Reply by mlesueur** [reply]
Thank you for your answers, I will update MOOSE and try to do a heap profiling. @lindsayad I use INS kernels and solve with NEWTON. I thought this was the way to use as little memory as possible and reduce CPU time. Is that correct?

**Reply by mlesueur** [reply]
I have updated moos. Now asm-ilu seems to be working, with a similar memory usage than FSP, at least at 1e6 DOFS. I haven't pushed it further yet. 
For the profiling, I haven't been able to use gperftools... Is it available with the latest moose version? So I got out the allocations from instruments instead, for both asm-ilu and fsp (https://we.tl/t-y8ndLzbOIv), for a small problem, 72488 DOFS. I don't really know how to read this type of files, but is this comprehensible for you?
Please let me know if I can do something else

**Reply by lindsayad** [reply]
> I use INS kernels and solve with NEWTON. I thought this was the way to use as little memory as possible and reduce CPU time. Is that correct?

That is correct

> Now asm-ilu seems to be working

So it's converging now without you making any other changes other than updating moose? Did you update your PETSc version?

> For the profiling, I haven't been able to use gperftools... Is it available with the latest moose version?

No. We used to distribute a package for it, but unfortunately we no longer do that. You would have to build that as well as libunwind most likely. If you are on Mac, it's my experience that gperftools doesn't work that well (maybe @fdkong has experienced otherwise), so instruments may be your best bet. I'll try looking at the traces next time I'm on my laptop

**Reply by fdkong** [reply]
> I have updated moos. Now asm-ilu seems to be working, with a similar memory usage than FSP, at least at 1e6 DOFS. I haven't pushed it further yet.
> For the profiling, I haven't been able to use gperftools... Is it available with the latest moose version? So I got out the allocations from instruments instead, for both asm-ilu and fsp (https://we.tl/t-y8ndLzbOIv), for a small problem, 72488 DOFS. I don't really know how to read this type of files, but is this comprehensible for you?
> Please let me know if I can do something else

@mlesueur Thanks for your results. In your test cases, FSP is better than ASM in terms of memory and compute time.

FSP, simulation time 42.562 s, memory 359 MB

ASM: simulation time 69.562 s,  memory 1030MB.

The reason is that you run simulations in serial and ASM is an ILU. ASM is competitive when running in parallel. But you should pick up whatever work better for you. 

I did not see unexpected memory allocations from your profiling results.

**Reply by mlesueur** [reply]
Thank you @fdkong for the feedback. Happy to know that the FSP looks like my best option. I guess the large memory usage is just normal when you are dealing with large meshes.
